### PR TITLE
Sass 302

### PR DESCRIPTION
**Pull request by BenCosfordHMRC**

### CODE REVIEW

File: app/controllers/IVFailureController.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                                    mcc: MessagesControllerComponents,
+                                    view: IVFailurePage,
+                                    implicit val ec: ExecutionContext) extends FrontendController(mcc) with I18nSupport with SessionDataHelper{
+
+  def show(journeyId: Option[String]): Action[AnyContent] = Action { implicit request =>
+
+    val sessionId: Option[String] = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session)).sessionId.map(_.value)
+
+    if(journeyId.isEmpty){
+      logger.warn(s"[IVFailureController][show] JourneyId from IV journey is empty. Defaulting journeyId for audit purposes." +
```
Comment (matt712) [inline app/controllers/IVFailureController.scala:44]: Is this the correct logging level? 
Don't we normally use info logging level

File: app/controllers/predicates/AuthorisedAction.scala [hunk 2]
```diff
@@ -18,64 +18,77 @@ package controllers.predicates
...
     } recover {
       case _: NoActiveSession =>
         logger.info(s"[AgentPredicate][authoriseAsAgent] - No active session. Redirecting to ${appConfig.signInUrl}")
         Redirect(appConfig.signInUrl) //TODO Check this is the correct location
+      case e: NoSuchElementException =>
+        if (e.getMessage.contains("Illegal confidence level")) {
+          logger.error("[AuthorisedAction][invokeBlock] User has invalid confidence level, routing user to IV uplift.")
+          Redirect(routes.IVUpliftController.initialiseJourney())
+        } else {
+          logger.error("[AuthorisedAction][invokeBlock] Received NoSuchElementException form auth when retrieving confidence level.")
```
Comment (matt712) [inline app/controllers/predicates/AuthorisedAction.scala:74]: This log is possibly misleading, one of the retrievals that aren't confidenceLevel could be missing resulting in this log.

File: app/controllers/testOnly/AdditionalSessionDataController.scala [hunk 3]
```diff
@@ -19,6 +19,7 @@ package controllers.testOnly
 import config.AppConfig
 import javax.inject.Inject
 import play.api.i18n.I18nSupport
+import play.api.libs.json.Json
```
Comment (matt712) [inline app/controllers/testOnly/AdditionalSessionDataController.scala]: Import not needed

**Review by matt712 — CHANGES_REQUESTED** [review]

File: app/controllers/IVFailureController.scala (hunk 1 above)
Comment (BenCosfordHMRC) [reply app/controllers/IVFailureController.scala:44]: I think whilst the parameter is a query parameter we should be warned if they don't actually give us the parameter.

File: app/controllers/predicates/AuthorisedAction.scala (hunk 2 above)
Comment (BenCosfordHMRC) [reply app/controllers/predicates/AuthorisedAction.scala]: Updated

**Pull request merged**